File: backend/schemas/api_spec.py
"""
Pydantic schemas for API request and response validation.
"""

from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime


class TagCreate(BaseModel):
    """Schema for creating a new tag."""
    name: str = Field(..., min_length=1, max_length=100)


class TagResponse(BaseModel):
    """Schema for tag response."""
    id: int
    name: str
    spec_count: int = 0
    
    class Config:
        from_attributes = True


class SpecCreate(BaseModel):
    """Schema for creating a new API spec."""
    name: str = Field(..., min_length=1, max_length=255)
    version: str = Field(..., min_length=1, max_length=50)
    provider: Optional[str] = Field(None, max_length=255)
    original_filename: Optional[str] = Field(None, max_length=255)
    original_format: Optional[str] = Field(None, pattern='^(yaml|json)$')
    original_content: str = Field(..., min_length=1)
    markdown_content: str = Field(..., min_length=1)
    token_count: Optional[int] = Field(None, ge=0)
    uploaded_by: Optional[str] = Field(None, max_length=255)
    file_size_bytes: Optional[int] = Field(None, ge=0)
    tags: Optional[List[str]] = Field(default_factory=list)


class SpecResponse(BaseModel):
    """Schema for API spec response (without full content)."""
    id: int
    name: str
    version: str
    provider: Optional[str]
    original_filename: Optional[str]
    original_format: Optional[str]
    token_count: Optional[int]
    uploaded_at: datetime
    uploaded_by: Optional[str]
    file_size_bytes: Optional[int]
    tags: List[str] = []
    
    class Config:
        from_attributes = True


class SpecDetail(SpecResponse):
    """Schema for detailed API spec response (includes full content)."""
    original_content: str
    markdown_content: str
    
    class Config:
        from_attributes = True


class SpecListResponse(BaseModel):
    """Schema for paginated list of specs."""
    total: int
    page: int
    page_size: int
    specs: List[SpecResponse]

File: frontend/src/types/api-spec.ts
/**
 * TypeScript types for API Spec marketplace
 * Matches backend Pydantic schemas
 */

export interface Tag {
  id: number
  name: string
  spec_count: number
}

export interface ApiSpec {
  id: number
  name: string
  version: string
  provider: string | null
  original_filename: string | null
  original_format: string | null
  token_count: number | null
  uploaded_at: string
  uploaded_by: string | null
  file_size_bytes: number | null
  tags: string[]
}

export interface SpecDetail extends ApiSpec {
  original_content: string
  markdown_content: string
}

export interface SpecListResponse {
  total: number
  page: number
  page_size: number
  specs: ApiSpec[]
}

export interface ChunkedSpec {
  manifest: string
  tags: Record<string, string>
  endpoints: Record<string, string>
  schemas: Record<string, string>
}

export interface ToolSchema {
  name: string
  description: string
  parameters: {
    type: string
    properties: Record<string, any>
    required?: string[]
  }
}

